Implement diacritics stripping and string normalization command

mierak opened the pull request:
Attempt to implement https://github.com/MusicPlayerDaemon/MPD/issues/2327

* Introduces new `stringnormalization` protocol command
* Implements first string normalization to strip diacritics when using `search` and the related commands

The implementation works, but I am unsure whether I took the right approach (and its my first go at CPP). The threading through of booleans is also getting pretty gnarly in there. So I am mostly looking for feedback at this point. We can also workshop the `stringnormalization` name.

mierak commented on src/song/StringFilter.hxx:
```diff
@@ -78,7 +78,7 @@ public:
 
 	bool GetFoldCase() const noexcept {
-		return fold_case;
+		return icu_compare;
```
Yup, I missed this one, thanks.

MaxKellermann commented:
Lots and lots of build failures:
```
../../test/TestStringFilter.cxx: In member function ‘virtual void StringFilterTest_ASCII_Test::TestBody()’:
../../test/TestStringFilter.cxx:23:82: error: no matching function for call to ‘StringFilter::StringFilter(<brace-enclosed initializer list>)’
   23 |         const StringFilter f{"needle", false, StringFilter::Position::FULL, false};
      |                                                                                  ^
```
And some bad coding style, uninitialized variables, .... sigh.

MaxKellermann commented:
Plus Windows build failures:
```
../../src/lib/icu/Compare.cxx:34:1: error: no declaration matches 'IcuCompare::IcuCompare(std::string_view)'
   34 | IcuCompare::IcuCompare(std::string_view _needle) noexcept
      | ^~~~~~~~~~
```
Same build failure when building without ICU.

When you fix a bad commit, don't add a fixup commit - instead, amend the existing known-bad commit.

mierak commented:
> Lots and lots of build failures:
> Same build failure when building without ICU.

Compiles for me now both with and without ICU. I have fixed the tests and added a simple test case for diacritic stripping both with and without fold case as well.

> When you fix a bad commit, don't add a fixup commit - instead, amend the existing known-bad commit.

I have squashed them into one since there was already more than one commit.

> Plus Windows build failures

Hopefully should be fine now, but I do not have means to test a windows build atm.

> bad coding style

Not much to go of off here, dev docs state to use spaces, I have tried to follow that as well as to copy the style elsewhere. Im going to need a bit more than "bad"

> uninitialized variables, .... sigh.

Which ones?

MaxKellermann commented on test/TestStringFilter.cxx:
```diff
@@ -130,17 +130,23 @@ TEST_F(StringFilterTest, Normalize)
+	EXPECT_TRUE(StringFilter("áéíóúýčďěňřšťžůåäöüàãâçêõîşûğăôơư", true, true, StringFilter::Position::FULL, false)
+            .Match("aeiouycdenrstzuaaouaaaceoisugaoou"));
+	EXPECT_TRUE(StringFilter("ÁÉÍÓÚÝČĎĚŇŘŠŤŽŮÅÄÖÜÀÃÂÇÊÕÎŞÛĞĂÔƠƯ", true, true, StringFilter::Position::FULL, false)
+            .Match("áéíóúýčďěňřšťžůåäöüàãâçêõîşûğăôơư"));
```
Bad indent

MaxKellermann reviewed (changes requested):
Just ... look at the diff and you see your indentation is all wrong.

mierak commented on src/lib/icu/Compare.hxx:
```diff
@@ -25,26 +25,54 @@ class IcuCompare {
 			? AllocatedString(src.needle)
 			: nullptr;
+        	fold_case = src.fold_case;
+        	strip_diacritics = src.strip_diacritics;
```
fixed

MaxKellermann commented on src/lib/icu/Compare.hxx:
```diff
@@ -25,26 +25,54 @@ class IcuCompare {
 			? AllocatedString(src.needle)
 			: nullptr;
+		fold_case = src.fold_case;
+		strip_diacritics = src.strip_diacritics;
 		return *this;
 	}
 
-	IcuCompare(IcuCompare &&) = default;
-	IcuCompare &operator=(IcuCompare &&) = default;
+	IcuCompare(IcuCompare &&src) noexcept
+		:needle(src
+			? AllocatedString(std::move(src.needle))
+			: nullptr),
+		 fold_case(src.fold_case),
+		 strip_diacritics(src.strip_diacritics)
+	{
+		src.needle = nullptr;
+		src.fold_case = false;
+		src.strip_diacritics = false;
+	}
+
+	IcuCompare &operator=(IcuCompare &&src) noexcept {
+		needle = src
+			? AllocatedString(std::move(src.needle))
+			: nullptr;
+		fold_case = src.fold_case;
+		strip_diacritics = src.strip_diacritics;
+
+		src.needle = nullptr;
+		src.fold_case = false;
+		src.strip_diacritics = false;
+		return *this;
+	}
```
What's the point of this change?

MaxKellermann replied:
That comment was wrong, sorry - I meant the copy ctor+operator, not copy+move ctor.
